### ISSUE TITLE
flash-attn is substantially outperformed by PyTorch's scaled_dot_product_attention (and installing from source is super slow)

### ISSUE THREAD

**Issue by netw0rkf10w** [Dao-AILab/flash-attention#827]
Hello,

I found that `flash-attn.flash_attn_func` is substantially outperformed by `torch.nn.functional.scaled_dot_product_attention` on an Nvidia A100 GPU, as shown in the results below.

Environment: newly installed `torch` (2.2.0) and `flash-attn` (2.5.3) via `pip`, CUDA 11.8.

```bash
[---------------------------------- FlashAttention ---------------------------------]
                                           |  sdp_attention  |  torch sdpa  |  flash 
1 threads: --------------------------------------------------------------------------
      [torch.float16, 12, 64, 256, 64]     |       519.4     |      98.7    |   364.1
      [torch.float16, 12, 64, 256, 128]    |       941.4     |     187.2    |   651.4
      [torch.float16, 16, 64, 256, 64]     |       641.2     |     128.5    |   447.3
      [torch.float16, 16, 64, 256, 128]    |      1242.4     |     246.3    |   859.7
      [torch.float16, 16, 64, 576, 64]     |      3435.5     |     695.7    |  1328.7
      [torch.float16, 16, 64, 576, 128]    |      6786.0     |    1387.5    |  2621.4
      [torch.float16, 16, 128, 256, 64]    |       736.6     |     234.7    |   845.1
      [torch.float16, 16, 128, 256, 128]   |      1435.4     |     459.0    |  1609.7
      [torch.float16, 16, 128, 784, 64]    |      6440.1     |    2056.5    |  3718.5
      [torch.float16, 16, 128, 784, 128]   |     12894.1     |    4085.1    |  7392.4
      [torch.bfloat16, 12, 64, 256, 64]    |       523.8     |      97.9    |   344.7
      [torch.bfloat16, 12, 64, 256, 128]   |       952.2     |     184.1    |   650.8
      [torch.bfloat16, 16, 64, 256, 64]    |       646.7     |     126.3    |   447.0
      [torch.bfloat16, 16, 64, 256, 128]   |      1256.1     |     240.9    |   858.7
      [torch.bfloat16, 16, 64, 576, 64]    |      3153.5     |     687.7    |  1318.2
      [torch.bfloat16, 16, 64, 576, 128]   |      6255.2     |    1372.2    |  2603.6
      [torch.bfloat16, 16, 128, 256, 64]   |       740.9     |     230.1    |   823.4
      [torch.bfloat16, 16, 128, 256, 128]  |      1442.1     |     446.7    |  1605.6
      [torch.bfloat16, 16, 128, 784, 64]   |      6478.1     |    2021.0    |  3688.0
      [torch.bfloat16, 16, 128, 784, 128]  |     12923.3     |    4023.6    |  7323.4

Times are in microseconds (us).
```

The second column (`sdp_attention`) is the math version. The grid variables are `[dtype, num_heads, head_dim, seq_len, batch_size]`. The input to the attention functions are just random vectors:
```python
q = torch.randn((B, H, L, D)).to('cuda').to(dtype)
k = torch.randn((B, H, L, D)).to('cuda').to(dtype)
v = torch.randn((B, H, L, D)).to('cuda').to(dtype)
```

The above values of `num_heads, head_dim, seq_len` correspond roughly to the configurations of ViT base, large, and huge (the head dimension of ViT-huge is 80, which is not supported by `flash-attn`, so I used `128` instead).

Could you tell what could possibly explain this huge difference in performance?

I tried building from source but gave up because it took too long (a few hours). In the past I had really installed flash attention from source a few times, but it only took a few minutes. I have `ninja` installed. I checked the CPU usage with `top`, a lot of cores were used, but only at a few percents. I also tried building the official ROCm fork of this repo on a server with MI250 cards, it only took a few minutes (the cores were at 100% most of the time). The ROCm fork seems to be a previous version of the current version, so maybe something changed between the two versions that made building slow?

**Comment by tridao** [comment]
Do you have a short script to reproduce the numbers?

**Comment by netw0rkf10w** [comment]
@tridao Here we go:

```python
import sys, os
import torch
import math
import torch.utils.benchmark as benchmark
from torch.nn.functional import scaled_dot_product_attention
import warnings

try:
    from flash_attn import flash_attn_func
    if torch.cuda.is_available():
        if torch.cuda.get_device_capability('cuda') >= (7, 5):
            HAS_FLASH = True
        else:
            warnings.warn("GPU not supported, Flash attention will not be used.")
            HAS_FLASH = False
    else:
        warnings.warn("Cuda not available, Flash attention will not be used.")
        HAS_FLASH = False
except ImportError:
    warnings.warn("Cannot import flash_attn, Flash attention will not be used.")
    HAS_FLASH = False

try:
    from torching.flash_attention import fa_triton
    if torch.cuda.is_available():
        if torch.cuda.get_device_capability('cuda') >= (7, 5):
            HAS_FLASH_TRITON = True
        else:
            warnings.warn("GPU not supported, Flash-triton will not be used.")
            HAS_FLASH_TRITON = False
    else:
        warnings.warn("Cuda not available, Flash-triton will not be used.")
        HAS_FLASH_TRITON = False
except ImportError:
    warnings.warn("Cannot import, Flash-triton will not be used.")
    HAS_FLASH_TRITON = False


# HAS_FLASH_TRITON = False

try:
    from torching.flash_attention import fa_triton_rocm
    if torch.cuda.is_available():
        if torch.cuda.get_device_capability('cuda') >= (9, 0):
            HAS_FLASH_TRITON_ROCM = True
        else:
            warnings.warn("GPU not supported, Flash-triton-ROCm will not be used.")
            HAS_FLASH_TRITON_ROCM = False
    else:
        warnings.warn("Cuda not available, Flash-triton-ROCm will not be used.")
        HAS_FLASH_TRITON_ROCM = False
except ImportError:
    warnings.warn("Cannot import, Flash-triton-ROCm will not be used.")
    HAS_FLASH_TRITON_ROCM = False


def sdp_attention(q, k, v, rel_pos=None):
    """
    q: (B, H, L, D)
    """
    # with torch.cuda.amp.autocast(enabled=True, dtype=q.dtype):
    scale = 1 / math.sqrt(q.size(-1))
    attn = (q @ k.transpose(-2, -1)) * scale # (B, H, L, L)
    if rel_pos is not None:
        attn = attn + rel_pos
    
    attn = attn.softmax(dim=-1) # (B, H, L, L)
    # attn = self.attn_drop(attn) 
    result = attn @ v
    # (B, H, L, L) x (B, H, L, D)  --> (B, H, L, D) --> (B, L, H, D)
    return result

def flash_attention(q, k, v):
    """
    shape: (B, H, L, D)
    """
    # flash attention expects (B, L, H, d)
    q = q.transpose(1, 2).contiguous()
    k = k.transpose(1, 2).contiguous()
    v = v.transpose(1, 2).contiguous()
    result = flash_attn_func(
                        q, k, v,
                        dropout_p=0,
                        softmax_scale=None,
                        causal=False,
                        return_attn_probs=False,
                        )
    return result.transpose(1, 2).contiguous()

def flash_attention_triton(q, k, v):
    """
    shape: (B, H, seqlen_q, D_HEAD)
    """
    output = fa_triton(q, k, v)
    return output

def flash_attention_triton_rocm(q, k, v):
    """
    shape: (B, H, seqlen_q, D_HEAD)
    """
    output, _ = fa_triton_rocm(q, k, v)
    return output
    

def ex():
    # (B, H, L, D)
    B = 5
    H = 4
    L = 10
    D = 16
    q = torch.randn((B, H, L, D)).to('cuda')
    k = torch.randn((B, H, L, D)).to('cuda')
    v = torch.randn((B, H, L, D)).to('cuda')

    # q = q.bfloat16(); k = k.bfloat16(); v = v.bfloat16()

    # bias = torch.randn((1, H, L, L))
    r1_fp32 = scaled_dot_product_attention(q, k, v)
    r1_fp16 = scaled_dot_product_attention(q.half(), k.half(), v.half())
    r1_bf16 = scaled_dot_product_attention(q.bfloat16(), k.bfloat16(), v.bfloat16())
    print(f"max diff r1_fp32 vs r1_fp16 = {(r1_fp32 - r1_fp16).abs().max()}")
    print(f"max diff r1_fp32 vs r1_bf16 = {(r1_fp32 - r1_bf16).abs().max()}")

    r2_fp32 = sdp_attention(q, k, v)
    r2_fp16 = sdp_attention(q.half(), k.half(), v.half())
    r2_bf16 = sdp_attention(q.bfloat16(), k.bfloat16(), v.bfloat16())
    print(f"max diff r1_fp32 vs r2_fp32 = {(r1_fp32 - r2_fp32).abs().max()}")
    print(f"max diff r1_fp32 vs r2_fp16 = {(r1_fp32 - r2_fp16).abs().max()}")
    print(f"max diff r1_fp32 vs r2_bf16 = {(r1_fp32 - r2_bf16).abs().max()}")

    if HAS_FLASH:
        # r3 = flash_attention(q, k, v)
        r3_fp16 = flash_attention(q.half(), k.half(), v.half())
        r3_bf16 = flash_attention(q.bfloat16(), k.bfloat16(), v.bfloat16())
        print(f"max diff r1_fp32 vs r3_fp16 = {(r1_fp32 - r3_fp16).abs().max()}")
        print(f"max diff r1_fp32 vs r3_bf16 = {(r1_fp32 - r3_bf16).abs().max()}")

    if HAS_FLASH_TRITON:
        r4_fp16 = flash_attention_triton(q.half(), k.half(), v.half())
        # r4_bf16 = flash_attention_triton(q.bfloat16(), k.bfloat16(), v.bfloat16())
        print(f"max diff r1_fp32 vs r4_fp16 = {(r1_fp32 - r4_fp16).abs().max()}")
        # print(f"max diff r1_fp32 vs r4_bf16 = {(r1_fp32 - r4_bf16).abs().max()}")

    if HAS_FLASH_TRITON_ROCM:
        r5_fp16 = flash_attention_triton_rocm(q.half(), k.half(), v.half())
        r5_bf16 = flash_attention_triton_rocm(q.bfloat16(), k.bfloat16(), v.bfloat16())
        print(f"max diff r1_fp32 vs r5_fp16 = {(r1_fp32 - r5_fp16).abs().max()}")
        print(f"max diff r1_fp32 vs r5_bf16 = {(r1_fp32 - r5_bf16).abs().max()}")


def main():
    # ViT base: embed_dim=768, num_heads=12, 
    #   patch=16, size=224: H=12, D=64, L=256 
    # ViT large: embed_dim=1024, num_heads=16:
    #   patch=14, size=224: H=16, D=64, L=256
    #   patch=16, size=384: H=16, D=64, L=576
    # ViT huge: embed_dim=1280, num_heads=16:
    #    patch=16, size=224: H=16, D=80, L=256
    #    patch=16, size=448: H=16, D=80, L=784

    batch_sizes = [64, 128]

    # Compare takes a list of measurements which we'll save in results.
    results = []

    dyptes = [torch.float16, torch.bfloat16]
    # heads = [12, 16]
    # dims = [64, 80, 128]
    # seq_lenghts = [256, 576, 784]

    # We replace 80 with 128 because the triton version does not support 80 
    HDL = [(12, 64, 256), (16, 64, 256), (16, 64, 576), (16, 128, 256), (16, 128, 784)]

    # HAS_FLASH_TRITON = False

    for dtype in dyptes:
        for H, D, L in HDL:
            for B in batch_sizes:
                q = torch.randn((B, H, L, D)).to('cuda').to(dtype)
                k = torch.randn((B, H, L, D)).to('cuda').to(dtype)
                v = torch.randn((B, H, L, D)).to('cuda').to(dtype)
                # label and sub_label are the rows
                # description is the column
                label = 'FlashAttention'
                sub_label = f'[{dtype}, {H}, {D}, {L}, {B}]'
                results.append(benchmark.Timer(
                    stmt='sdp_attention(q, k, v)',
                    setup='from __main__ import sdp_attention',
                    globals={'q': q, 'k': k, 'v': v},
                    # num_threads=num_threads,
                    label=label,
                    sub_label=sub_label,
                    description='sdp_attention',
                ).blocked_autorange(min_run_time=1))
                results.append(benchmark.Timer(
                    stmt='scaled_dot_product_attention(q, k, v)',
                    setup='from torch.nn.functional import scaled_dot_product_attention',
                    globals={'q': q, 'k': k, 'v': v},
                    # num_threads=num_threads,
                    label=label,
                    sub_label=sub_label,
                    description='torch sdpa',
                ).blocked_autorange(min_run_time=1))
                if dtype != torch.float32:
                    if HAS_FLASH:
                        results.append(benchmark.Timer(
                            stmt='flash_attention(q, k, v)',
                            setup='from __main__ import flash_attention',
                            globals={'q': q, 'k': k, 'v': v},
                            # num_threads=num_threads,
                            label=label,
                            sub_label=sub_label,
                            description='flash',
                        ).blocked_autorange(min_run_time=1))
                    if HAS_FLASH_TRITON:
                        results.append(benchmark.Timer(
                            stmt='flash_attention_triton(q, k, v)',
                            setup='from __main__ import flash_attention_triton',
                            globals={'q': q, 'k': k, 'v': v},
                            # num_threads=num_threads,
                            label=label,
                            sub_label=sub_label,
                            description='triton',
                        ).blocked_autorange(min_run_time=1))
                    if HAS_FLASH_TRITON_ROCM:
                        results.append(benchmark.Timer(
                            stmt='flash_attention_triton_rocm(q, k, v)',
                            setup='from __main__ import flash_attention_triton_rocm',
                            globals={'q': q, 'k': k, 'v': v},
                            # num_threads=num_threads,
                            label=label,
                            sub_label=sub_label,
                            description='triton-rocm',
                        ).blocked_autorange(min_run_time=1))


    compare = benchmark.Compare(results)
    compare.print()


if __name__ == '__main__':
    main()
    # ex()
```

**Comment by tridao** [comment]
```
    q = q.transpose(1, 2).contiguous()
    k = k.transpose(1, 2).contiguous()
    v = v.transpose(1, 2).contiguous()
    return result.transpose(1, 2).contiguous()
```
These are hurting your performance since they involve memory copy. You don't need contiguous.

**Comment by netw0rkf10w** [comment]
> ```
>     q = q.transpose(1, 2).contiguous()
>     k = k.transpose(1, 2).contiguous()
>     v = v.transpose(1, 2).contiguous()
>     return result.transpose(1, 2).contiguous()
> ```
> 
> These are hurting your performance since they involve memory copy. You don't need contiguous.

Ahhh good catch! Let me try again!

**Comment by netw0rkf10w** [comment]
I confirm that `flash-attn` is actually the fastest:

```bash
[--------------------------------------- FlashAttention ---------------------------------------]
                                           |  sdp_attention  |  torch sdpa  |  flash   |  triton
1 threads: -------------------------------------------------------------------------------------
      [torch.float16, 12, 64, 256, 64]     |       488.6     |      98.7    |   120.6  |   264.1
      [torch.float16, 12, 64, 256, 128]    |       941.5     |     186.9    |   178.1  |   202.7
      [torch.float16, 16, 64, 256, 64]     |       641.8     |     128.5    |   121.9  |   293.2
      [torch.float16, 16, 64, 256, 128]    |      1242.1     |     246.1    |   236.1  |   268.6
      [torch.float16, 16, 64, 576, 64]     |      3387.4     |     702.6    |   646.3  |   754.2
      [torch.float16, 16, 64, 576, 128]    |      6713.9     |    1391.3    |  1281.7  |  1237.3
      [torch.float16, 16, 128, 256, 64]    |       737.7     |     234.5    |   233.3  |   417.4
      [torch.float16, 16, 128, 256, 128]   |      1434.6     |     458.0    |   414.6  |   530.0
      [torch.float16, 16, 128, 784, 64]    |      6443.8     |    2061.2    |  1933.8  |  2419.8
      [torch.float16, 16, 128, 784, 128]   |     12868.0     |    4080.9    |  3845.1  |  4554.8
      [torch.bfloat16, 12, 64, 256, 64]    |       494.3     |      97.7    |    93.3  |        
      [torch.bfloat16, 12, 64, 256, 128]   |       951.0     |     184.5    |   176.3  |        
      [torch.bfloat16, 16, 64, 256, 64]    |       647.0     |     126.0    |   119.2  |        
      [torch.bfloat16, 16, 64, 256, 128]   |      1255.5     |     240.2    |   230.5  |        
      [torch.bfloat16, 16, 64, 576, 64]    |      3154.0     |     689.0    |   634.3  |        
      [torch.bfloat16, 16, 64, 576, 128]   |      6262.9     |    1366.8    |  1261.1  |        
      [torch.bfloat16, 16, 128, 256, 64]   |       740.9     |     230.6    |   209.7  |        
      [torch.bfloat16, 16, 128, 256, 128]  |      1437.1     |     446.0    |   404.9  |        
      [torch.bfloat16, 16, 128, 784, 64]   |      6478.9     |    2026.0    |  1896.0  |        
      [torch.bfloat16, 16, 128, 784, 128]  |     12927.1     |    4007.4    |  3756.1  |        

Times are in microseconds (us).

```

PyTorch uses `flash-attn` behind the scene so it's not surprising that its performance is close (the difference probably comes from the pre-processing overhead). `flash-attn` is only slightly outperformed by `triton` in a single case: `[torch.float16, 16, 64, 576, 128]`, and outperformed by `torch`'s SDPA for `[torch.float16, 12, 64, 256, 64]`; I'm not sure what's special about these instances...

Thanks again, @tridao! Your work is amazing!

**Comment by vadimkantorov** [comment]
Triton version is considerably slower though...

**Comment by Huiimin5** [comment]
In my environment, flash attention is about 3x slower than torch sdpa. Do you have any idea why?

<img width="870" alt="image" src="https://github.com/user-attachments/assets/79e1c584-bb06-4a94-8f47-0c32d1aef5eb" />
